storage: combine spans when adding to the timestamp cache

petermattis opened the pull request:
```
name                  old time/op    new time/op    delta
KVInsert1_SQL-32         504µs ± 2%     508µs ± 2%  +0.94%  (p=0.008 n=19+20)
KVInsert10_SQL-32        956µs ± 2%     934µs ± 2%  -2.34%  (p=0.000 n=20+20)
KVInsert100_SQL-32      5.05ms ± 2%    4.81ms ± 2%  -4.70%  (p=0.000 n=20+18)
KVInsert1000_SQL-32     48.5ms ± 5%    45.0ms ± 3%  -7.09%  (p=0.000 n=19+19)
KVInsert10000_SQL-32     651ms ±10%     647ms ±13%    ~     (p=0.828 n=18+20)

name                  old allocs/op  new allocs/op  delta
KVInsert1_SQL-32           304 ± 0%       302 ± 0%  -0.66%  (p=0.000 n=19+20)
KVInsert10_SQL-32          754 ± 0%       717 ± 0%  -4.96%  (p=0.000 n=19+16)
KVInsert100_SQL-32       5.04k ± 0%     4.65k ± 0%  -7.83%  (p=0.000 n=20+20)
KVInsert1000_SQL-32      48.5k ± 2%     44.4k ± 1%  -8.46%  (p=0.000 n=20+20)
KVInsert10000_SQL-32      565k ± 9%      525k ± 9%  -7.06%  (p=0.002 n=20+20)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6413)

<!-- Reviewable:end -->

petermattis committed:
storage: avoid the interval tree when there are no conflicts

Instead of directly adding the spans from a request to the timestamp
cache interval trees, they are now gathered into a `cacheRequest`
structure and placed in a btree keyed by timestamp. When applying the
timestamp cache to an incoming request, before retrieving the max
read/write timestamps from the cache we expand any `cacheRequests` that
are newer than request being executed. The common case is that the
executing request is newer than any previous request, and we thus
usually avoid inserting spans into the interval trees.

name                            old time/op    new time/op    delta
TrackChoices1_Cockroach-32         691µs ± 5%     620µs ± 1%  -10.21%  (p=0.000 n=20+19)
TrackChoices10_Cockroach-32        217µs ± 5%     161µs ± 6%  -25.57%  (p=0.000 n=20+19)
TrackChoices100_Cockroach-32       152µs ± 4%     102µs ± 4%  -32.71%  (p=0.000 n=20+18)
TrackChoices1000_Cockroach-32      143µs ± 5%     103µs ± 8%  -27.97%  (p=0.000 n=19+19)
InsertDistinct1_Cockroach-32      2.64ms ± 8%    2.11ms ± 9%  -19.88%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32     2.20ms ± 5%    1.49ms ± 4%  -32.18%  (p=0.000 n=20+20)
InsertDistinct100_Cockroach-32    2.59ms ± 4%    2.05ms ± 6%  -20.54%  (p=0.000 n=20+19)
KVInsert1_SQL-32                   544µs ± 4%     541µs ± 3%     ~     (p=0.687 n=19+20)
KVInsert10_SQL-32                  963µs ± 3%     950µs ± 4%   -1.39%  (p=0.006 n=18+18)
KVInsert100_SQL-32                4.17ms ± 6%    3.83ms ± 8%   -8.25%  (p=0.000 n=20+20)
KVInsert1000_SQL-32               34.4ms ± 8%    30.9ms ± 9%  -10.00%  (p=0.000 n=18+20)
KVInsert10000_SQL-32               385ms ± 8%     355ms ±15%   -7.97%  (p=0.001 n=17+19)

name                            old allocs/op  new allocs/op  delta
TrackChoices1_Cockroach-32           395 ± 0%       384 ± 0%   -2.79%  (p=0.000 n=20+20)
TrackChoices10_Cockroach-32          148 ± 0%       136 ± 0%   -8.11%  (p=0.000 n=20+20)
TrackChoices100_Cockroach-32         120 ± 0%       107 ± 0%  -10.83%  (p=0.000 n=17+18)
TrackChoices1000_Cockroach-32        117 ± 0%       104 ± 0%  -11.11%  (p=0.000 n=20+18)
InsertDistinct1_Cockroach-32       1.90k ± 0%     1.76k ± 0%   -7.36%  (p=0.000 n=17+17)
InsertDistinct10_Cockroach-32      2.02k ± 0%     1.92k ± 0%   -4.95%  (p=0.000 n=20+19)
InsertDistinct100_Cockroach-32     4.72k ± 0%     4.62k ± 0%   -2.25%  (p=0.000 n=20+18)
KVInsert1_SQL-32                     282 ± 0%       279 ± 0%   -0.99%  (p=0.000 n=20+20)
KVInsert10_SQL-32                    656 ± 0%       622 ± 0%   -5.25%  (p=0.000 n=15+20)
KVInsert100_SQL-32                 4.13k ± 0%     3.74k ± 0%   -9.56%  (p=0.000 n=20+20)
KVInsert1000_SQL-32                38.8k ± 0%     34.7k ± 0%  -10.63%  (p=0.000 n=20+18)
KVInsert10000_SQL-32                416k ± 0%      372k ± 3%  -10.63%  (p=0.000 n=18+20)

petermattis commented:
@tschottdorf This is intended for discussion. See also #6412. Similar to that change, perhaps we should use some heuristics to combine spans.

petermattis commented:
@tschottdorf Do you have opinions about this change?

petermattis commented:
Here are updated numbers where old is #6412:

```
name                            old time/op    new time/op    delta
InsertDistinct1_Cockroach-32      2.55ms ± 5%    1.99ms ± 4%  -21.85%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32     2.19ms ± 5%    1.50ms ± 4%  -31.31%  (p=0.000 n=20+20)
InsertDistinct100_Cockroach-32    2.56ms ± 6%    2.04ms ± 7%  -20.08%  (p=0.000 n=20+18)
KVInsert1_SQL-32                   503µs ± 2%     495µs ± 2%   -1.51%  (p=0.001 n=19+19)
KVInsert10_SQL-32                  934µs ± 2%     898µs ± 2%   -3.79%  (p=0.000 n=17+20)
KVInsert100_SQL-32                4.79ms ± 4%    4.51ms ± 6%   -5.86%  (p=0.000 n=20+20)
KVInsert1000_SQL-32               42.6ms ± 7%    38.5ms ± 7%   -9.77%  (p=0.000 n=19+18)
KVInsert10000_SQL-32               494ms ±10%     436ms ±10%  -11.77%  (p=0.000 n=18+19)

name                            old allocs/op  new allocs/op  delta
InsertDistinct1_Cockroach-32       2.07k ± 4%     1.95k ± 3%   -6.04%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32      2.23k ± 3%     2.11k ± 3%   -5.22%  (p=0.000 n=19+20)
InsertDistinct100_Cockroach-32     4.95k ± 2%     4.88k ± 2%   -1.41%  (p=0.000 n=20+17)
KVInsert1_SQL-32                     294 ± 0%       292 ± 0%   -0.68%  (p=0.000 n=20+19)
KVInsert10_SQL-32                    705 ± 0%       668 ± 0%   -5.25%  (p=0.000 n=20+20)
KVInsert100_SQL-32                 4.62k ± 0%     4.23k ± 0%   -8.54%  (p=0.000 n=20+20)
KVInsert1000_SQL-32                43.8k ± 0%     39.8k ± 0%   -9.14%  (p=0.000 n=17+20)
KVInsert10000_SQL-32                459k ± 0%      419k ± 0%   -8.75%  (p=0.000 n=16+19)
```

So definitely a performance win, but are there correctness or other issues from this change?

bdarnell commented:
I think this would be correct, but I would expect performance problems from the wider spans in certain situations, just like in #6412. Since the timestamp cache is mainly there for read/write conflicts, I think a benchmark of transactions that do a bunch of reads (same pattern as InsertDistinct) followed by at least one write would be the best way to expose that performance difference.

tschottdorf commented:
Like @bdarnell pointed out, we don't at all exercise the downside of this change in the above benchmarks. The cost we pay for this optimization in practice are transaction restarts under contention. With the exponential backoff we have now, these create a lot of variance (the bank benchmark is the only one we have that actually attempts this, I think) and so we haven't really looked at them much.

A workload which for some reason writes to the beginning and end of a key range in one batch will have all concurrent writers in between restart, which will make some workloads which should be completely conflict-free contended. Even worse, it could lead to starvation. Picture a reader busy-looping a read of keys `a` and `z`, while everyone else is writing random, non-intersecting keys in between. Currently, that's completely conflict-free. With the change, the reader bombs the whole range repeatedly with its read timestamp. All in-flight writers will lay down an intent but have to restart; but when they come back, the timestamp cache again instructs them to restart. They lay down (update) their intent, but when they come back... you get the idea. With this change, we're losing the behavior that laying down an intent forces a concurrent reader into conflict resolution - you're introducing artificial "readers" here which actually never see the intents.

To benchmark changes like this, I think what we might want to consider is "retryless contention" - i.e. we run transactions that give up when they catch a restart or intent error (since a lot of the variance came from internal retry loops). That should give us some measure of how "contended" a workload is (without measuring how much effort a transaction would have to do to succeed).

But, in this case, we have to figure out the starvation scenario outlined above first.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[roachpb/data.go, line 931 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KGwlJ9-3fttIyn2eTUd:-KGwlJ9-3fttIyn2eTUe:-2044422802) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7f1802437795834d06394b6306766a65f02b5d/roachpb/data.go#L931)):*
We should try to avoid relying on `nil != []byte{}` if we can (or `[nil,a)` union `[b, c)` ends up `[b,c)`). I believe all should be well if you just drop this special case (though ultimately the signature of this method suggests that it shouldn't be called on an uninitialized Span at all: `[nil,nil)` could technically stand for a point request to `KeyMin`, but that's not what you mean).
Also, should unit test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KGwnTX4o6m7Ib6Lpaax:355380824)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

petermattis commented:
In most workloads, there isn't conflict. Similar to the `CommandQueue` optimization, I think we can make the `TimestampCache` add covering spans that are split on conflict. For the timestamp cache, conflict is not just when a request encounters a covering span, but also when the new request has a lower timestamp than the covering span. For the `TrackChoices` benchmark, we would never actually split the covering spans as the transaction timestamps are always increasing. I also benchmarked `TrackChoices` against this PR (old) vs a hack to never add to the timestamp cache (new):

```
name                          old time/op    new time/op    delta
TrackChoices1_Cockroach-8        391µs ± 1%     354µs ± 1%  -9.44%          (p=0.008 n=5+5)
TrackChoices10_Cockroach-8       111µs ± 1%     107µs ± 0%  -3.89%          (p=0.008 n=5+5)
TrackChoices100_Cockroach-8     74.8µs ± 1%    72.9µs ± 2%  -2.53%          (p=0.008 n=5+5)
TrackChoices1000_Cockroach-8    77.6µs ± 3%    76.0µs ± 1%    ~             (p=0.222 n=5+5)
```

As expected, completely skipping the timestamp cache is faster for small batch sizes. But at large batch size other effects take over. 

And here is how this PR benchmarks against master:

```
name                          old time/op    new time/op    delta
TrackChoices1_Cockroach-8        445µs ± 2%     391µs ± 1%  -12.24%  (p=0.008 n=5+5)
TrackChoices10_Cockroach-8       143µs ± 1%     111µs ± 1%  -22.32%  (p=0.008 n=5+5)
TrackChoices100_Cockroach-8      112µs ± 0%      75µs ± 1%  -33.07%  (p=0.008 n=5+5)
TrackChoices1000_Cockroach-8     108µs ± 2%      78µs ± 3%  -28.46%  (p=0.008 n=5+5)
```

petermattis commented:
This change has been completely revamped. I need to regenerate the performance numbers, but I expect them to be equivalent or better. And there is likely some additional cleanup and testing work to do. Would be good to get some eyes on the general approach, though. PTAL.

tschottdorf commented:
I like this more than trying to split the covering span on a conflict (because I think an optimization that works best for sorted input is brittle - add an index write somewhere and suddenly the sorting is off). Curious to see the performance impact. Running InsertDistinct should also be interesting - it should profit here as well where previous attempts wouldn't.
Some more comments in the code explaining the BTree would be good.
This code will do very well for insert-only workloads. What about insert workloads with non-overlapping reads? Right now, the writes would break all the reads' entries, though they wouldn't overlap. We could make the splitting conditional on there being a possible conflict, but I doubt it's worth it.

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This change has been completely revamped. I need to regenerate the performance numbers, but I expect them to be equivalent or better. And there is likely some additional cleanup and testing work to do. Would be good to get some eyes on the general approach, though. PTAL.</details>


---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 78 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfQRBtzQDnWt7h2tB7:-KHfQRBuvW50tGsxytxN:476100858) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L78)):*
> ```Go
> 	lowWater, latest roachpb.Timestamp
> 
> 	requests   *btree.BTree
> ```

Add a comment here.

---

*[storage/timestamp_cache.go, line 319 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfRYS1bacUAdVoCwaW:-KHfRYS1bacUAdVoCwaX:-1401782070) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L319)):*
> ```Go
> 
> 	for {
> 		minItem := tc.requests.Min()
> ```

Can Ascend do this more efficiently (by not descending to the min from the root in each iteration of this loop)?

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfS3vieTXBQKtac2Bm:-49695297) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
> ```Go
> 		// log.Infof("removing request: %s: %d/%d/%v\n",
> 		// 	minReq.timestamp, len(minReq.reads), len(minReq.writes), minReq.txn.Key != nil)
> 		tc.requests.DeleteMin()
> ```

In effect, if a request hasn't been expanded while it's in the window, we'll never expand it, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHfTaUQHvQnI3mcskQ5:795485963)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfWgEJ2ok4hb5-qBux:-1313161633) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> In effect, if a request hasn't been expanded while it's in the window, we'll never expand it, right?</details>
Oh wait, I think I got that wrong - we never inserted it in the first place. Aren't we losing something here? We want to allow inserts in the far past (well, wanted, anyway) for interactive SQL sessions. They shouldn't be forced higher than the low water mark - this was previously achieved by evicting based on `evictionSizeThreshold`, but I think is now broken (unless you clash with spans frequently, in which case the "old" spans stay around after the threshold).
Maybe interactive sessions simply ought to do their inserts early. Or do you see a way out to achieve the same effect without doing more work?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The latest commit contains updated benchmarks. `InsertDistinct` saw a nice improvement. Yes, we could make the splitting more intelligent. I'll add a comment to the code. We don't currently have a benchmark which shows a benefit to doing that.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I like this more than trying to split the covering span on a conflict (because I think an optimization that works best for sorted input is brittle - add an index write somewhere and suddenly the sorting is off). Curious to see the performance impact. Running InsertDistinct should also be interesting - it should profit here as well where previous attempts wouldn't.
> Some more comments in the code explaining the BTree would be good.
> This code will do very well for insert-only workloads. What about insert workloads with non-overlapping reads? Right now, the writes would break all the reads' entries, though they wouldn't overlap. We could make the splitting conditional on there being a possible conflict, but I doubt it's worth it.</details>


---

Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 78 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfQRBtzQDnWt7h2tB7:-KHfX-ZUn6KEEaFTeZP6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L78)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add a comment here.</details>
Done.

---

*[storage/timestamp_cache.go, line 319 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfRYS1bacUAdVoCwaW:-KHfX9JzxShAqysFJPWE:1087874630) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L319)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Can Ascend do this more efficiently (by not descending to the min from the root in each iteration of this loop)?</details>
Yeah, we could probably use `AscendLessThan`. Let me add a TODO.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfXDxsjP8xlrY6T0If:1798441257) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> In effect, if a request hasn't been expanded while it's in the window, we'll never expand it, right?</details>
Correct.

---

*[roachpb/data.go, line 931 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KGwlJ9-3fttIyn2eTUd:-KHfXHA8o95mTuVa--r0:689104986) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7f1802437795834d06394b6306766a65f02b5d/roachpb/data.go#L931)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> We should try to avoid relying on `nil != []byte{}` if we can (or `[nil,a)` union `[b, c)` ends up `[b,c)`). I believe all should be well if you just drop this special case (though ultimately the signature of this method suggests that it shouldn't be called on an uninitialized Span at all: `[nil,nil)` could technically stand for a point request to `KeyMin`, but that's not what you mean).
> Also, should unit test.</details>
This has been reverted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHfXIq-RSNpVeuWHv1b:1678923129)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfYVx-pCD6F0ts_y9s:-158938515) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Correct.</details>
Err, my "Correct" was in response to your first comment. Can you expand on the problem you're seeing here? I'm not following you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfYyF3_Nj9J3FO_1VQ:1851741402) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Err, my "Correct" was in response to your first comment. Can you expand on the problem you're seeing here? I'm not following you.</details>
Sure. As long as the timestamp cache is empty, an insert far in the past will hit the low water mark. More precisely, naively (and a while ago) elements were eagerly thrown out of the cache, so that you would (roughly) never get a response smaller than `now-10s` out of the cache. We changed that to keep old elements around (governed by `evictionSizeThreshold`), the intention being that as long as there isn't much activity on the cache, it would be able to return values smaller than `now-10s` (i.e. the low water mark wouldn't just rise up eagerly). With your change, you *do* go back to evicting eagerly from the btree, and you do change the watermark. So my >10s SQL transaction will have its timestamp pushed if I try to insert something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfaRGsiLITMaSJ8B6c:-1952318913) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Sure. As long as the timestamp cache is empty, an insert far in the past will hit the low water mark. More precisely, naively (and a while ago) elements were eagerly thrown out of the cache, so that you would (roughly) never get a response smaller than `now-10s` out of the cache. We changed that to keep old elements around (governed by `evictionSizeThreshold`), the intention being that as long as there isn't much activity on the cache, it would be able to return values smaller than `now-10s` (i.e. the low water mark wouldn't just rise up eagerly). With your change, you *do* go back to evicting eagerly from the btree, and you do change the watermark. So my >10s SQL transaction will have its timestamp pushed if I try to insert something.</details>
Would it be sufficient to "expand" a request being evicted if `TimestampCache.Len() < evictionSizeThreshold`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHgUKC6ZThOB4QeyL5h:1744579018) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Would it be sufficient to "expand" a request being evicted if `TimestampCache.Len() < evictionSizeThreshold`?</details>
I now track how many spans are present in the `requests` queue and stop evicting requests when the number drops below `evictionSizeThreshold`. We don't seem to have a test for this behavior and I couldn't reproduce a failure without this change in manual testing. Do you have a suggestion for how to test? Should I file an issue to write such a test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The latest commit contains updated benchmarks. `InsertDistinct` saw a nice improvement. Yes, we could make the splitting more intelligent. I'll add a comment to the code. We don't currently have a benchmark which shows a benefit to doing that.</details>


---

Reviewed 2 of 3 files at r5, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/timestamp_cache.go, line 345 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHrXapinY3IzJyPcDFe:-KHrXapj0-51WVPf5sBg:-882757401) ([raw file](https://github.com/cockroachdb/cockroach/blob/a398b2eaa6874c0fdc8a387d1e6c2b3253c8d729/storage/timestamp_cache.go#L345)):*
> ```Go
> 
> 	// Evict requests as long as the number of cached spans (both in the requests
> 	// queue and the interval caches) is larger that the eviction threshold.
> ```

s/that/than/

---

*[storage/timestamp_cache.go, line 465 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHrY3uQVeNpfwe0AuNb:-KHrY3uQVeNpfwe0AuNc:1069945688) ([raw file](https://github.com/cockroachdb/cockroach/blob/a398b2eaa6874c0fdc8a387d1e6c2b3253c8d729/storage/timestamp_cache.go#L465)):*
> ```Go
> 		dest.lowWater = tc.lowWater
> 		dest.latest = tc.latest
> 		dest.requests = btree.New(64)
> ```

Define a constant for this repeated 64.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHrYJZEP0JxJ7d4uq-g:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 345 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHrXapinY3IzJyPcDFe:-KHtXc_YNiLLlXcl3b_C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a398b2eaa6874c0fdc8a387d1e6c2b3253c8d729/storage/timestamp_cache.go#L345)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/that/than/</details>
Done.

---

*[storage/timestamp_cache.go, line 465 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHrY3uQVeNpfwe0AuNb:-KHtXdsklKXfP1EIzNgC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a398b2eaa6874c0fdc8a387d1e6c2b3253c8d729/storage/timestamp_cache.go#L465)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Define a constant for this repeated 64.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf Want to give this another look?

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> @tschottdorf Want to give this another look?</details>


---

Reviewed 2 of 3 files at r5.
Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1032 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtxJlEYtfEJZ2kM6Pt:-KHtxJlEYtfEJZ2kM6Pu:-1304770356) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/replica.go#L1032)):*
> ```Go
> 					// DeleteRange adds to the write timestamp cache to prevent
> 					// subsequent writes from rewriting history.
> 					cr.writes = append(cr.writes, args.Header())
> ```

`s/args.Header()/header`

---

*[storage/replica.go, line 1071 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtxU-JLWR1KHT3di2V:-KHtxU-JLWR1KHT3di2W:1426506064) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/replica.go#L1071)):*
> ```Go
> 	defer r.mu.Unlock()
> 
> 	if ba.Txn != nil {
> ```

Is this `if ba.Txn != nil ... get timestamp` logic used elsewhere? It might be deserving of a method on `*roachpb.BatchRequest`.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHu0UJrdnXXowb2EqYg:1097890027) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I now track how many spans are present in the `requests` queue and stop evicting requests when the number drops below `evictionSizeThreshold`. We don't seem to have a test for this behavior and I couldn't reproduce a failure without this change in manual testing. Do you have a suggestion for how to test? Should I file an issue to write such a test?</details>
Huh, this wasn't failing `TestTimestampCacheNoEviction`?

---

*[storage/timestamp_cache.go, line 41 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHty974pqPNloxIca5e:-KHty9753-S__vszWi-g:-1506062416) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L41)):*
> ```Go
> 	defaultEvictionSizeThreshold = 512
> 
> 	// Max entries in each btree node. Not yet tuned.
> ```

Maybe add a TODO to tune.

---

*[storage/timestamp_cache.go, line 61 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtySsm09MGga7dEHE6:-KHtySsm09MGga7dEHE7:711783346) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L61)):*
> ```Go
> }
> 
> func (cr *cacheRequest) Less(other btree.Item) bool {
> ```

Is this fitting an interface needed for the `btree`? If so, it might be nice to have a comment, even if one isnt needed for govet.

---

*[storage/timestamp_cache.go, line 63 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtytr7lalnJldcYdE_:-KHtytr7lalnJldcYdEa:-489315415) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L63)):*
> ```Go
> func (cr *cacheRequest) Less(other btree.Item) bool {
> 	otherReq := other.(*cacheRequest)
> 	if cr.timestamp.Less(otherReq.timestamp) {
> ```

Nit: I'd be curious how much logic could be cleaned up with a `roachpb.Timestamp.Compare` method.

---

*[storage/timestamp_cache.go, line 69 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtz4WBW3SNRSJ9Q4I7:-KHtz4WBW3SNRSJ9Q4I8:2061386844) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L69)):*
> ```Go
> 		return false
> 	}
> 	// Fallback to pointer comparison as a tie-breaker. This allows multiple
> ```

Pointer comparison? Was this comment from before uniqueID was used?

---

*[storage/timestamp_cache.go, line 341 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHu-Aq_tS5Az35xjGql:-KHu-Aq_tS5Az35xjGqm:1145679489) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L341)):*
> ```Go
> 
> 	// Bump the latest timestamp and evict any requests that are now too old.
> 	if tc.latest.Less(req.timestamp) {
> ```

Consider `tc.latest.Forward`

---

*[storage/timestamp_cache.go, line 395 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHu19Y7on1jue3qCJl0:-KHu19Y7on1jue3qCJl1:1411314594) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L395)):*
> ```Go
> 		tc.reqSpans -= req.numSpans()
> 		for _, sp := range req.reads {
> 			tc.Add(sp.Key, sp.EndKey, req.timestamp, req.txnID, true /* readTSCache */)
> ```

Is this the only place `tc.Add` is called from? If so, maybe unexport the method. I know it doesnt really do anything because everything is within the `storage` package anyway, but it will make the external interface of the cache a little more clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHu1s22mf7yHTH6VgDM:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1032 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtxJlEYtfEJZ2kM6Pt:-KHu5QHfFksXdKTF6Zg0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/replica.go#L1032)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/args.Header()/header`</details>
Done.

---

*[storage/replica.go, line 1071 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtxU-JLWR1KHT3di2V:-KHu5efh_0iKwyJdIjcH:43466846) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/replica.go#L1071)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Is this `if ba.Txn != nil ... get timestamp` logic used elsewhere? It might be deserving of a method on `*roachpb.BatchRequest`.</details>
Yeah, let me see about doing that in a separate change. Not sure what to call that method, though.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHu6anm9zbTfWozC3U3:2028863860) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Huh, this wasn't failing `TestTimestampCacheNoEviction`?</details>
`TestTimestampCacheNoEviction` doesn't use `AddRequest`. 

---

*[storage/timestamp_cache.go, line 41 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHty974pqPNloxIca5e:-KHu5qalqUoLZpR-9cPr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Maybe add a TODO to tune.</details>
Done.

---

*[storage/timestamp_cache.go, line 61 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtySsm09MGga7dEHE6:-KHu5x3BP5gUkhOBW893:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L61)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Is this fitting an interface needed for the `btree`? If so, it might be nice to have a comment, even if one isnt needed for govet.</details>
Done.

---

*[storage/timestamp_cache.go, line 63 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtytr7lalnJldcYdE_:-KHu6-tpJIrKNoRi9cIZ:676283418) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L63)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Nit: I'd be curious how much logic could be cleaned up with a `roachpb.Timestamp.Compare` method.</details>
It would be slightly fewer lines of code, though I'm not sure if it would be any simpler.

---

*[storage/timestamp_cache.go, line 69 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHtz4WBW3SNRSJ9Q4I7:-KHu68_bu3javmwnoWS-:-1114453878) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Pointer comparison? Was this comment from before uniqueID was used?</details>
Yes, pointer comparison was originally used before I discovered it doesn't work. Fixed the comment.

---

*[storage/timestamp_cache.go, line 341 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHu-Aq_tS5Az35xjGql:-KHu6PI50qEflh26GsrW:743268223) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L341)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Consider `tc.latest.Forward`</details>
Done. And also changed in `TimestampCache.Add` where this was copied from. 

---

*[storage/timestamp_cache.go, line 395 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHu19Y7on1jue3qCJl0:-KHu6nCTu5FcU0H3ZuD_:-919257627) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef12da5a1ab41bcf65404c80ec99be5496c2cc4/storage/timestamp_cache.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Is this the only place `tc.Add` is called from? If so, maybe unexport the method. I know it doesnt really do anything because everything is within the `storage` package anyway, but it will make the external interface of the cache a little more clear.</details>
Yes (except for tests). `TimestampCache` itself doesn't need to be exported. I'll fix this in a follow-on change as it will be a mechanical renaming. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod test update?

<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> :lgtm:</details>


---

Reviewed 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHfoFr0M44BHS_no9_M:2024745240) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `TestTimestampCacheNoEviction` doesn't use `AddRequest`. </details>
Haven't looked at the test, but looks like it should be updated (to fail before your latest addition)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHuU9d1I60TY2BtEhTi:-196855595)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/timestamp_cache.go, line 330 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-KHfS3vhbk6w8zc08mA0:-KHuVYjKWNMN2ZIIIXax:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22519780a59aa220d65f2fd9bd9a81d9ac9a0507/storage/timestamp_cache.go#L330)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Haven't looked at the test, but looks like it should be updated (to fail before your latest addition)?</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> :lgtm: mod test update?</details>


---

Reviewed 2 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6413#-:-KHuW-t1F8kMZ9GMkTmV:1417088721)*
<!-- Sent from Reviewable.io -->
